File: cpp/Prefix Sums/sum_between_range.cpp
#include <vector>

class SumBetweenRange {
public:
    SumBetweenRange(std::vector<int>& nums) {
        if (!nums.empty()) {
            prefixSum.push_back(nums[0]);
            for (int i = 1; i < nums.size(); i++) {
                prefixSum.push_back(prefixSum.back() + nums[i]);
            }
        }
    }

    int sumRange(int i, int j) {
        if (i == 0) {
            return prefixSum[j];
        }
        return prefixSum[j] - prefixSum[i - 1];
    }

private:
    std::vector<int> prefixSum;
};

File: java/Prefix Sums/SumBetweenRange.java
public class SumBetweenRange {
    int[] prefixSum;

    public SumBetweenRange(int[] nums) {
        this.prefixSum = new int[nums.length];
        this.prefixSum[0] = nums[0];
        for (int i = 1; i < nums.length; i++) {
            this.prefixSum[i] = this.prefixSum[i - 1] + nums[i];
        }
    }

    public int sumRange(int i, int j) {
        if (i == 0) {
            return this.prefixSum[j];
        }
        return this.prefixSum[j] - this.prefixSum[i - 1];
    }
}
